kernel/src/process/signal/pauser.rs

```diff
@@ -1,101 +1,33 @@
-    time::wait::WaitTimeout,
+    prelude::*, process::posix_thread::PosixThreadExt, thread::Thread, time::wait::WaitTimeout,
 };
 
-/// A `Pauser` allows pausing the execution of the current thread until certain conditions are reached.
-///
-/// Behind the scene, `Pauser` is implemented with [`Waiter`] and [`WaitQueue`].
-/// But unlike its [`Waiter`] relatives, `Pauser` is aware of POSIX signals:
-/// if a thread paused by a `Pauser` receives a signal, then the thread will resume its execution.
-///
-/// Another key difference is that `Pauser` combines the two roles of [`Waiter`] and [`WaitQueue`]
-/// into one. Both putting the current thread to sleep and waking it up can be done through the
-/// same `Pauser` object, using its `pause`- and `resume`-family methods.
-///
-/// [`Waiter`]: ostd::sync::Waiter
-///
-/// # Example
-///
-/// Here is how the current thread can be put to sleep with a `Pauser`.
-///
-/// ```no_run
-/// let pauser = Pauser::new(SigMask::new_full());
-/// // Pause the execution of the current thread until a user-given condition is met
-/// // or the current thread is interrupted by a signal.
-/// let res = pauser.pause_until(|| {
-///     if cond() {
-///         Some(())
-///     } else {
-///         None
-///     }
-/// });
-/// match res {
-///     Ok(_) => {
-///         // The user-given condition is met...
-///     }
-///     Err(EINTR) => {
-///         // A signal is received...
-///     }
-///     _ => unreachable!()
-/// }
-/// ```
-///
-/// Let's assume that another thread has access to the same object of `Arc<Pauser>`.
-/// Then, this second thread can resume the execution of the first thread
-/// even when `cond()` does not return `true`.
-///
-/// ```no_run
-/// pauser.resume_all();
-/// ```
-pub struct Pauser {
-    wait_queue: WaitQueue,
-    sig_mask: SigMask,
+/// A trait to provide inner pause implementations for [`Waiter`] and [`WaitQueue`]`.
+trait PauseInner {
+    fn do_pause<F, R>(&self, cond: F, timeout: Option<&Duration>) -> Result<R>
+    where
+        F: FnMut() -> Option<R>;
 }
 
-impl Pauser {
-    /// Creates a new `Pauser`.
-    ///
-    /// The `Pauser` can be interrupted by all signals
-    /// except that are blocked by current thread.
-    pub fn new() -> Arc<Self> {
-        Self::new_with_mask(SigMask::new_empty())
-    }
-
-    /// Creates a new `Pauser` with specified `sig_mask`.
-    ///
-    /// The `Pauser` will ignore signals that are in `sig_mask`
-    /// or blocked by current thread.
-    pub fn new_with_mask(sig_mask: SigMask) -> Arc<Self> {
-        let wait_queue = WaitQueue::new();
-        Arc::new(Self {
-            wait_queue,
-            sig_mask,
-        })
-    }
-
+/// The `Pause` trait provides the pause ability for [`Waiter`] and [`WaitQueue`]`,
+/// which can allow the waiting process of them be interrupted by POSIX signals,
+/// i.e. if a paused thread receives a signal, then the thread will resume its execution.
+#[allow(private_bounds)]
```

Comment:
This seems better to use [sealed trait](https://predr.ag/blog/definitive-guide-to-sealed-traits-in-rust/)?

Something like:
```rust
mod internal {
    pub trait PauseInner {}
}
```